lumicks/pylake/kymotracker/detail/denoising.py

```diff
@@ -0,0 +1,412 @@
+            ]
+
+        detail_coefficients = [f1 - f2 for f1, f2 in zip(filtered_imgs, filtered_imgs[1:])]
+        remainder = filtered_imgs[-1]
+
+        return detail_coefficients, remainder
+
+    def _reconstruct_image(self, detail_coefficients, remainder, stabilize):
+        """Reconstructs image from wavelet decomposition.
+
+        Parameters
+        ----------
+        detail_coefficients : List[np.ndarray]
+            List of images representing the detail coefficients for each wavelet.
+        remainder : np.ndarray
+            Image containing the remainder (what's left after adding the detail layers).
+        stabilize : bool
+            Indicate that the detail coefficients were stabilized and this transform
+            must be inverted.
+        """
+        output_image = sum(detail_coefficients + [remainder])
+        if not stabilize:
+            return output_image
+
+        return inverse_variance_stabilizing_transform(output_image, self._coefficients[0])
+
+    def process_image(self, image, false_detection_rate=0.1, verbose=False):
+        """Process an image with MS-VST.
+
+        Parameters
+        ----------
+        image : np.ndarray
+            Image array
+        false_detection_rate : float
+            Represents the False Detection Rate when it comes to pixels with significant
+            signal. The probability of erroneously detecting spots in a spot-free homogeneous
+            noise is upper bounded by this value.
+        verbose : bool
+            Show extra output
+        """
+        image = np.array(image, dtype=float)  # integer arrays lose quality when convolved
+        detail_coefficients, remainder = self._calculate_wavelets(image, stabilize=True)
+
+        if verbose:
+            for d, stdev in zip(detail_coefficients, self._stdev):
+                print(f"practical: {np.std(d.flatten())}, theoretical: {stdev}")
+
+        # Determine significant coefficients. At this point, the distribution is approximately
+        # normal and the standard deviation is known by construction.
+        significant = [
+            determine_significant(d, stdev, false_detection_rate=false_detection_rate)
+            for d, stdev in zip(detail_coefficients, self._stdev)
+        ]
+
+        # Filter the insignificant coefficients
+        detail_coefficients = [d * sig for d, sig in zip(detail_coefficients, significant)]
+
+        # Inverse transform
+        output_image = self._reconstruct_image(detail_coefficients, remainder, stabilize=True)
+
+        # Enforce positivity. Error model states that negative values don't make sense. Yet they
+        # can still occur because of the overcompleteness of the IUWT and the fact that we
+        # modified values in the detail coefficients directly.
+        output_image[output_image < 0] = 0
+
+        return output_image, significant
+
+    def regularized_reconstruction(
```

Comment:
nitpick, would be nice to have a verb here. `filter_regularized()`? `reconstruct_regularized()`?